situp-plugins/peer-forwarder/src/main/java/com/amazon/situp/plugins/processor/peerforwarder/PeerForwarder.java

```diff
@@ -0,0 +1,140 @@
+        }
+        for (final Record<ExportTraceServiceRequest> record: records) {
+            for (final ResourceSpans rs: record.getData().getResourceSpansList()) {
+                final List<Map.Entry<String, ResourceSpans>> rsBatch = PeerForwarderUtils.splitByTrace(rs);
+                for (final Map.Entry<String, ResourceSpans> entry: rsBatch) {
+                    final String traceId = entry.getKey();
+                    final ResourceSpans newRS = entry.getValue();
+                    final String peerIp = getHostByConsistentHashing(traceId);
+                    peerGroupedRS.get(peerIp).add(newRS);
+                }
+            }
+        }
+
+        // Send ExportTraceServiceRequest to designated peer
+        final List<Record<ExportTraceServiceRequest>> results = new ArrayList<>();
+        for (final String peerIp: peerIps) {
+            final TraceServiceGrpc.TraceServiceBlockingStub client = peerClients.getOrDefault(peerIp, null);
+            ExportTraceServiceRequest.Builder currRequestBuilder = ExportTraceServiceRequest.newBuilder();
+            int currSpansCount = 0;
+            for (final ResourceSpans rs: peerGroupedRS.get(peerIp)) {
+                final int rsSize = PeerForwarderUtils.getResourceSpansSize(rs);
+                if (currSpansCount >= peerForwarderConfig.getMaxNumSpansPerRequest()) {
+                    final ExportTraceServiceRequest currRequest = currRequestBuilder.build();
+                    processRequest(client, currRequest, results);
+                    currRequestBuilder = ExportTraceServiceRequest.newBuilder();
+                    currSpansCount = 0;
+                }
+                currRequestBuilder.addResourceSpans(rs);
+                currSpansCount += rsSize;
+            }
+            // last request
+            final ExportTraceServiceRequest currRequest = currRequestBuilder.build();
+            processRequest(client, currRequest, results);
+        }
+        return results;
+    }
+
+    /**
+     * Forward request to the peer address if client is given, otherwise push the request to local buffer.
+     */
+    private void processRequest(final TraceServiceGrpc.TraceServiceBlockingStub client,
+                                final ExportTraceServiceRequest request,
+                                final List<Record<ExportTraceServiceRequest>> localBuffer) {
+        if (client != null) {
+            try {
+                client.export(request);
+            } catch (Exception e) {
+                LOG.error(String.format("Failed to forward the request:\n%s\n", request.toString()));
+                localBuffer.add(new Record<>(request));
+            }
+        } else {
+            localBuffer.add(new Record<>(request));
+        }
+    }
+
+    private String getHostByConsistentHashing(final String traceId) {
+        // TODO: better consistent hashing algorithm, e.g. ring hashing
+        return peerIps.get(traceId.hashCode() % peerIps.size());
```

Comment:
`hashCode()` can return negative integers - need absolute value here

Reply:
Yes. Thanks for the catch. I thought mod will yield a positive residue. But this seems not the case in Java.